From rostedt@goodmis.org Tue Jun 19 04:41:17 2007
Return-Path: <rostedt@goodmis.org>
X-Spam-Checker-Version: SpamAssassin 3.1.7-deb (2006-10-05) on debian
X-Spam-Level: 
X-Spam-Status: No, score=0.0 required=5.0 tests=AWL autolearn=unavailable 
	version=3.1.7-deb
Received: from ms-smtp-01.nyroc.rr.com (ms-smtp-01.nyroc.rr.com
	[24.24.2.55]) by mail.tglx.de (Postfix) with ESMTP id 4C19265C3EC for
	<tglx@linutronix.de>; Tue, 19 Jun 2007 04:41:17 +0200 (CEST)
Received: from [192.168.23.10] (cpe-24-94-51-176.stny.res.rr.com
	[24.94.51.176]) by ms-smtp-01.nyroc.rr.com (8.13.6/8.13.6) with ESMTP id
	l5J2f9l0013971; Mon, 18 Jun 2007 22:41:10 -0400 (EDT)
Subject: [PATCH RT] Don't call mcount from vsyscall_fn's
From: Steven Rostedt <rostedt@goodmis.org>
To: Ingo Molnar <mingo@elte.hu>
Cc: Thomas Gleixner <tglx@linutronix.de>, LKML <linux-kernel@vger.kernel.org>, RT <linux-rt-users@vger.kernel.org>
Content-Type: text/plain
Date: Mon, 18 Jun 2007 22:41:09 -0400
Message-Id: <1182220869.15228.10.camel@localhost.localdomain>
Mime-Version: 1.0
X-Mailer: Evolution 2.6.3 
X-Virus-Scanned: Symantec AntiVirus Scan Engine
X-Evolution-Source: imap://tglx%40linutronix.de@localhost:8993/
Content-Transfer-Encoding: 8bit

This bit me in the butt.

I couldn't understand why my init app was segfaulting, with a kernel
address, but a user RIP and RSP.  Well, the RIP I think was bogus, but
the kernel address was always the start of "mcount".  Looking deeper, I
printed out what was in the RSP (even though it was a user stack).  It
ended up showing me that the calling address was from the VDSO area.
Looking even further, I found the offending culprit, which was
vread_hpet.

Looking at the assembly dump, I saw the vread_hpet was calling mcount,
but I could not see it in the code. Nor could I see it in hpet.i (-E
option of compiling).

Well, I guess Ingo is a magician when it comes to compiler tricks, and
has the mcount being called by "every!!" function, unless you add the
"notrace" option.

This patch adds the notrace to vsyscall_fn, so that we don't have user
land apps calling mcount and crashing!

Signed-off-by: Steven Rostedt <rostedt@goodmis.org>

---
 include/asm-x86_64/vsyscall.h |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: linux-2.6.23.1-rt5/include/asm-x86_64/vsyscall.h
===================================================================
--- linux-2.6.23.1-rt5.orig/include/asm-x86_64/vsyscall.h
+++ linux-2.6.23.1-rt5/include/asm-x86_64/vsyscall.h
@@ -24,7 +24,7 @@ enum vsyscall_num {
 	((unused, __section__ (".vsyscall_gtod_data"),aligned(16)))
 #define __section_vsyscall_clock __attribute__ \
 	((unused, __section__ (".vsyscall_clock"),aligned(16)))
-#define __vsyscall_fn __attribute__ ((unused,__section__(".vsyscall_fn")))
+#define __vsyscall_fn __attribute__ ((unused,__section__(".vsyscall_fn"))) notrace
 
 #define VGETCPU_RDTSCP	1
 #define VGETCPU_LSL	2
